File: hdf-generators/go/ruby_escape.go
package generators

import "strings"

// EscapeQuotes wraps a string as a Ruby string literal.
//
// Strategy:
//   - Both single and double quotes → %q() wrapper
//   - Only single quotes → double-quoted with backslash escaping
//   - Otherwise → single-quoted with backslash escaping
func EscapeQuotes(s string) string {
	hasSingle := strings.Contains(s, "'")
	hasDouble := strings.Contains(s, `"`)

	if hasSingle && hasDouble {
		// %q() — escape backslashes before ) so Ruby doesn't treat \) as escaped delimiter
		escaped := strings.ReplaceAll(s, `\)`, `\\)`)
		return "%q(" + escaped + ")"
	}

	if hasSingle {
		// Double-quoted: escape backslashes, then double quotes
		escaped := strings.ReplaceAll(s, `\`, `\\`)
		escaped = strings.ReplaceAll(escaped, `"`, `\"`)
		return `"` + escaped + `"`
	}

	// Single-quoted: escape backslashes, then single quotes
	escaped := strings.ReplaceAll(s, `\`, `\\`)
	escaped = strings.ReplaceAll(escaped, `'`, `\'`)
	return "'" + escaped + "'"
}

File: hdf-generators/src/ruby-escape.ts
/**
 * Escape a string for use as a Ruby string literal.
 *
 * Strategy:
 * - If the string contains both single and double quotes → %q() wrapper
 * - If the string contains only single quotes → double-quoted with escaping
 * - Otherwise → single-quoted with escaping
 *
 * Ported from ts-inspec-objects escapeQuotes(), cross-referenced with
 * inspec-parser's RubyRebuilder.
 */
export function escapeQuotes(s: string): string {
  const hasSingle = s.includes("'");
  const hasDouble = s.includes('"');

  if (hasSingle && hasDouble) {
    // %q() — escape backslashes before ) so Ruby doesn't treat \) as escaped delimiter
    return `%q(${s.replace(/\\\)/g, '\\\\)')})`;
  }

  if (hasSingle) {
    // Double-quoted: escape backslashes, then double quotes
    return `"${s.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`;
  }

  // Single-quoted: escape backslashes, then single quotes
  return `'${s.replace(/\\/g, '\\\\').replace(/'/g, "\\'")}'`;
}
